sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -400,13 +400,24 @@ object PushProjectionThroughUnion extends Rule[LogicalPlan] with PredicateHelper
     // Push down deterministic projection through UNION ALL
     case p @ Project(projectList, Union(children)) =>
       assert(children.nonEmpty)
-      if (projectList.forall(_.deterministic)) {
-        val newFirstChild = Project(projectList, children.head)
+      val (deterministicList, nonDeterministic) = projectList.partition(_.deterministic)
+
+      if (deterministicList.nonEmpty) {
+        val newFirstChild = Project(deterministicList, children.head)
         val newOtherChildren = children.tail.map { child =>
           val rewrites = buildRewrites(children.head, child)
-          Project(projectList.map(pushToRight(_, rewrites)), child)
+          Project(deterministicList.map(pushToRight(_, rewrites)), child)
```

Comment:
do we push `a + 1` to union children? or just `a`?

Reply:
if we push a + 1 to union, just a + 1.
seems this lack of test cases. let's add this test cases. thanks.